### PR TITLE
Fix: (rewrites) incorrect parsing of destination query

### LINKED ISSUE #26609: «Rewrite mechanism» transforms array parameters of query of destination link into strings
### What version of Next.js are you using?

10.2.2

### What version of Node.js are you using?

14.17.1

### What browser are you using?

Safari

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

I'm trying to use Next.js' Rewrite feature. But when I put an array in a query part of a destination link string (in next.config.js), a context, that I receive in `getServerSideProps`, contains only first parameter of the array.

Here is my next.config.js

```
module.exports = {
...
  async rewrites() {
    return [
      {
        source: '/projects',
        destination: '/?tags=project&tags=проект',
      },
    ]
  },
...
}
```

When I open /projects in a browser, only first value of `tags` parameter is passed into `context.query` of `getServerSideProps`:

```
{
    tags: 'project'
}
```

### Expected Behavior

`context.query` in `getServerSideProps` must contain all query values.

```
{
    tags: ['project', 'проект']
}
```

### To Reproduce

* Create a new next.js project with the following next.config.js:

```
module.exports = {
...
  async rewrites() {
    return [
      {
        source: '/projects',
        destination: '/?tags=project&tags=проект',
      },
    ]
  },
...
}
```

* Create a page 'index.tsx' with getServerSideProps:
```
export const getServerSideProps: GetServerSideProps<any> = async (context) => {
    ....
    console.log(context.query);
    ....
};
```

* Open `/projects` page in a browser and check the server-side terminal

----

Actually, I've figured out why this happens and i'm ready to create PR. But before, I need to make sure that I understood the problem.

Here is the code, that causes the bug (**packages/next/next-server/lib/router/utils/prepare-destination.ts**):

```
export default function prepareDestination(
  destination: string,
  params: Params,
  query: ParsedUrlQuery,
  appendParamsToQuery: boolean
) {
...
  // update any params in query values
  for (const [key, strOrArray] of Object.entries(destQuery)) {
    let value = Array.isArray(strOrArray) ? strOrArray[0] : strOrArray
    if (value) {
      // the value needs to start with a forward-slash to be compiled
      // correctly
      value = compileNonPath(value, params)
    }
    destQuery[key] = value
  }
...
```

As you can see, it takes only a first item of the query array. I believe we must use` .map `here

### PR DESCRIPTION
**Pull request by lsndr**
fixes #26609

## Bug

- [x]  Related issues linked using fixes #number
- [x]   Integration tests added

## Documentation / Examples

- [x]  Make sure the linting passes

### CODE REVIEW

**Comment by lsndr** [comment]
Azure tests have failed for the reasons, not related to my commits

**Review by styfle — DISMISSED** [review]
Thanks!

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| buildDuration | 11.5s | 11.7s | ⚠️ +186ms |
| buildDurationCached | 2.8s | 2.8s | -20ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +87 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.039 | 2.092 | ⚠️ +0.05 |
| / avg req/sec | 1226.17 | 1195.27 | ⚠️ -30.9 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.169 | 1.171 | 0 |
| /error-in-render avg req/sec | 2137.72 | 2134.3 | ⚠️ -3.42 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| 359.HASH.js gzip | 3.09 kB | 3.09 kB | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 20.6 kB | 20.6 kB | ✓ |
| webpack-HASH.js gzip | 1.49 kB | 1.49 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 803 B | 803 B | ✓ |
| _error-HASH.js gzip | 3.18 kB | 3.18 kB | ✓ |
| amp-HASH.js gzip | 526 B | 526 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| hooks-HASH.js gzip | 903 B | 903 B | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.65 kB | 1.65 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.42 kB | 8.42 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 390 B | 390 B | ✓ |
| Overall change | 390 B | 390 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| index.html gzip | 523 B | 523 B | ✓ |
| link.html gzip | 537 B | 537 B | ✓ |
| withRouter.html gzip | 515 B | 515 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| buildDuration | 10.2s | 10.3s | ⚠️ +84ms |
| buildDurationCached | 4.3s | 4.1s | -216ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +87 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.059 | 2.035 | -0.02 |
| / avg req/sec | 1214 | 1228.7 | +14.7 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.143 | 1.171 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 2187.2 | 2134.12 | ⚠️ -53.08 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| 14.HASH.js gzip | 3.11 kB | 3.11 kB | ✓ |
| 677f882d2ed8..HASH.js gzip | 13.9 kB | 13.9 kB | ✓ |
| framework.HASH.js gzip | 41.8 kB | 41.8 kB | ✓ |
| main-HASH.js gzip | 7.81 kB | 7.81 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 791 B | 791 B | ✓ |
| _error-HASH.js gzip | 3.83 kB | 3.83 kB | ✓ |
| amp-HASH.js gzip | 531 B | 531 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| hooks-HASH.js gzip | 910 B | 910 B | ✓ |
| index-HASH.js gzip | 227 B | 227 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 295 B | 295 B | ✓ |
| withRouter-HASH.js gzip | 292 B | 292 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.98 kB | 8.98 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 418 B | 418 B | ✓ |
| Overall change | 418 B | 418 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| index.html gzip | 569 B | 569 B | ✓ |
| link.html gzip | 581 B | 581 B | ✓ |
| withRouter.html gzip | 561 B | 561 B | ✓ |
| Overall change | 1.71 kB | 1.71 kB | ✓ |

</details>

</details>
Commit: 1da4fde93c94501b6d4529fd288f424251b8709b

File: test/integration/rewrites-destination-query-array/pages/index.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import Link from 'next/link'
```
Comment (styfle) [inline test/integration/rewrites-destination-query-array/pages/index.js:1]: Remove this unused import so lint passes

**Review by ijjk — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| buildDuration | 13.5s | 13.1s | -411ms |
| buildDurationCached | 3.3s | 3.2s | -87ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +87 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.493 | 2.534 | ⚠️ +0.04 |
| / avg req/sec | 1002.97 | 986.42 | ⚠️ -16.55 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.618 | 1.552 | -0.07 |
| /error-in-render avg req/sec | 1545.18 | 1610.6 | +65.42 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| 359.HASH.js gzip | 3.09 kB | 3.09 kB | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 20.6 kB | 20.6 kB | ✓ |
| webpack-HASH.js gzip | 1.49 kB | 1.49 kB | ✓ |
| Overall change | 67.2 kB | 67.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 803 B | 803 B | ✓ |
| _error-HASH.js gzip | 3.18 kB | 3.18 kB | ✓ |
| amp-HASH.js gzip | 526 B | 526 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| hooks-HASH.js gzip | 903 B | 903 B | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.65 kB | 1.65 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.42 kB | 8.42 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 390 B | 390 B | ✓ |
| Overall change | 390 B | 390 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| index.html gzip | 523 B | 523 B | ✓ |
| link.html gzip | 537 B | 537 B | ✓ |
| withRouter.html gzip | 515 B | 515 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| buildDuration | 11.7s | 11.4s | -264ms |
| buildDurationCached | 4.7s | 4.6s | -165ms |
| nodeModulesSize | 49.3 MB | 49.3 MB | ⚠️ +87 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.497 | 2.546 | ⚠️ +0.05 |
| / avg req/sec | 1001.29 | 981.85 | ⚠️ -19.44 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.649 | 1.603 | -0.05 |
| /error-in-render avg req/sec | 1516.17 | 1559.76 | +43.59 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| 14.HASH.js gzip | 3.11 kB | 3.11 kB | ✓ |
| 677f882d2ed8..HASH.js gzip | 13.9 kB | 13.9 kB | ✓ |
| framework.HASH.js gzip | 41.8 kB | 41.8 kB | ✓ |
| main-HASH.js gzip | 7.81 kB | 7.81 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _app-HASH.js gzip | 791 B | 791 B | ✓ |
| _error-HASH.js gzip | 3.83 kB | 3.83 kB | ✓ |
| amp-HASH.js gzip | 531 B | 531 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| hooks-HASH.js gzip | 910 B | 910 B | ✓ |
| index-HASH.js gzip | 227 B | 227 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 295 B | 295 B | ✓ |
| withRouter-HASH.js gzip | 292 B | 292 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 8.98 kB | 8.98 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| _buildManifest.js gzip | 418 B | 418 B | ✓ |
| Overall change | 418 B | 418 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | lsndr/next.js dev-fix | Change |
| - | - | - | - |
| index.html gzip | 569 B | 569 B | ✓ |
| link.html gzip | 581 B | 581 B | ✓ |
| withRouter.html gzip | 561 B | 561 B | ✓ |
| Overall change | 1.71 kB | 1.71 kB | ✓ |

</details>

</details>
Commit: 03142be89cead26ce461cbcfb53d3189dffaee68

**Pull request merged**